src/fmt.zig

```diff
@@ -42,6 +44,73 @@ pub fn resetWith(fmt: *Formatter, ast: IR) void {
     fmt.ast = ast;
 }
 
+/// Formats all roc files in the specified path.
+/// Handles both single files and directories
+/// Returns the number of files formatted.
+pub fn formatPath(gpa: std.mem.Allocator, base_dir: std.fs.Dir, path: []const u8) !usize {
+    var count: usize = 0;
+    // First try as a directory.
+    if (base_dir.openDir(path, .{ .iterate = true })) |const_dir| {
+        var dir = const_dir;
+        defer dir.close();
+        // Walk is recursive.
+        var walker = try dir.walk(gpa);
+        defer walker.deinit();
+        while (try walker.next()) |entry| {
+            if (entry.kind == .file) {
+                if (try formatFilePath(gpa, entry.dir, entry.basename)) {
+                    count += 1;
+                }
+            }
+        }
+    } else |_| {
+        if (try formatFilePath(gpa, base_dir, path)) {
+            count += 1;
+        }
+    }
+
+    return count;
+}
+
+fn formatFilePath(gpa: std.mem.Allocator, base_dir: std.fs.Dir, path: []const u8) !bool {
+    // Format any ".roc" files.
+    if (std.mem.eql(u8, std.fs.path.extension(path), ".roc")) {
+        const input_file = try base_dir.openFile(path, .{ .mode = .read_only });
+        defer input_file.close();
+
+        const contents = try input_file.reader().readAllAlloc(gpa, std.math.maxInt(usize));
+        defer gpa.free(contents);
+
+        var module_env = base.ModuleEnv.init(gpa);
+        defer module_env.deinit();
+
+        var parse_ast = parse(&module_env, contents);
+        defer parse_ast.deinit();
+        if (parse_ast.errors.len > 0) {
+            // TODO: pretty print the parse failures.
+            const stderr = std.io.getStdErr().writer();
+            try stderr.print("Failed to parse '{s}' for formatting.\nErrors:\n", .{path});
+            for (parse_ast.errors) |err| {
+                try stderr.print("\t{s}\n", .{@tagName(err.tag)});
+            }
+            fatal("\n", .{});
```

Comment:
Based on [what I see in the old compiler](https://github.com/roc-lang/roc/blob/327647a6161b96d06f6524ee393ab675c2fc1335/crates/cli/src/format.rs#L202), this "exit on malformed code" seems to mostly match our current behavior, but I don't know if it's desired to fail if there are literally any parsing problems. What if there's a single unfinished function body, but everything else is valid?

Did I miss a discussion on whether this was what we wanted? It's fine for now, but we'll want to make sure we align here at some point.

Reply:
Long term the goal is that most errors will be recoverable and get formatted. For now, that is not anywhere near the case. So we just fail.